include/dlaf/matrix/hdf5.h

```diff
@@ -0,0 +1,274 @@
+
+  const H5::DataSpace& ds_file = dataset.getSpace();
+
+  for (const auto ij : common::iterate_range2d(dist.localNrTiles())) {
+    auto map_ds_tile = [&](auto&& tile) {
+      const GlobalElementIndex tl = dist.globalElementIndex(dist.globalTileIndex(ij), {0, 0});
+
+      const hsize_t file_counts[3] = {
+          to_sizet(tile.size().cols()),
+          to_sizet(tile.size().rows()),
+          internal::hdf5_datatype<T>::dims,
+      };
+      const hsize_t file_offsets[3] = {
+          to_sizet(tl.col()),
+          to_sizet(tl.row()),
+          0,
+      };
+      ds_file.selectHyperslab(H5S_SELECT_SET, file_counts, file_offsets);
+
+      const hsize_t memory_dims[3] = {
+          to_sizet(tile.size().cols()),
+          to_sizet(tile.ld()),
+          internal::hdf5_datatype<T>::dims,
+      };
+      H5::DataSpace ds_memory(3, memory_dims);
+
+      const hsize_t memory_counts[3] = {
+          to_sizet(tile.size().cols()),
+          to_sizet(tile.size().rows()),
+          internal::hdf5_datatype<T>::dims,
+      };
+      const hsize_t memory_offsets[3] = {0, 0, 0};
+      ds_memory.selectHyperslab(H5S_SELECT_SET, memory_counts, memory_offsets);
+
+      dataset_op(tile.ptr(), internal::hdf5_datatype<T>::type, ds_memory, ds_file);
+    };
+
+    tt::sync_wait(get_tile(ij) | di::transform(di::Policy<Backend::MC>(), std::move(map_ds_tile)));
+  }
+}
+
+template <class T>
+void from_dataset(const H5::DataSet& dataset, dlaf::Matrix<T, Device::CPU>& matrix) {
+  auto tile_rw = [&matrix](LocalTileIndex ij) { return matrix.readwrite(ij); };
+  auto dataset_read = [&dataset](auto&&... args) {
+    dataset.read(std::forward<decltype(args)>(args)...);
+  };
+  for_each_local_map_with_dataset<T>(matrix.distribution(), dataset, std::move(tile_rw),
+                                     std::move(dataset_read));
+}
+
+template <class T>
+void to_dataset(dlaf::Matrix<const T, Device::CPU>& matrix, const H5::DataSet& dataset) {
+  auto tile_ro = [&matrix](LocalTileIndex ij) { return matrix.read(ij); };
+  auto dataset_write = [&dataset](auto&&... args) {
+    dataset.write(std::forward<decltype(args)>(args)...);
+  };
+  for_each_local_map_with_dataset<T>(matrix.distribution(), dataset, std::move(tile_ro),
+                                     std::move(dataset_write));
+}
+
+class FileHDF5 final {
+public:
+  /// File access modes:
+  /// - readonly, the file will be opened (if should already exist)
+  /// - readwrite, the file will be created (if should not exist)
+  enum class FileMode {
+    readonly,
+    readwrite,
+  };
+
+  /// Create/open a local file.
+  ///
+  /// @p filepath filepath where the file to be opened is or where it will be created
+  /// @p mode file access mode (see FileMode for more details)
+  ///
+  /// @post if mode == READWRITE, being a local file it will not support parallel-write
+  FileHDF5(const std::string& filepath, const FileMode& mode) {
+    file_ = H5::H5File(filepath, mode2flags(mode));
```

Comment:
Let me take the chance also to show you the error messages, since in this PR we don't "filter" them and we let HDF5 output its own error messages.

This is a run on RO filesystem
```bash
ialberto@daint106:/project/csstaff/ialberto> DLAF_DEBUG_DUMP_TRISOLVER_DATA=1 srun -n 1 -c72 ~/workspace/dla-future.hdf5/build-mc/miniapp/miniapp_eigensolver --matrix-size 10240 --block-size 512 --nwarmups=0 --nruns 1
[0]
HDF5-DIAG: Error detected in HDF5 (1.14.0) MPI-process 0:
  #000: /tmp/ialberto/spack-stage/spack-stage-hdf5-1.14.0-6nwv5jrxawmg6fqv7z35kbhocugdu47y/spack-src/src/H5F.c line 836 in H5Fopen(): unable to synchronously open file
    major: File accessibility
    minor: Unable to open file
  #001: /tmp/ialberto/spack-stage/spack-stage-hdf5-1.14.0-6nwv5jrxawmg6fqv7z35kbhocugdu47y/spack-src/src/H5F.c line 796 in H5F__open_api_common(): unable to open file
    major: File accessibility
    minor: Unable to open file
  #002: /tmp/ialberto/spack-stage/spack-stage-hdf5-1.14.0-6nwv5jrxawmg6fqv7z35kbhocugdu47y/spack-src/src/H5VLcallback.c line 3863 in H5VL_file_open(): open failed
    major: Virtual Object Layer
    minor: Can't open object
  #003: /tmp/ialberto/spack-stage/spack-stage-hdf5-1.14.0-6nwv5jrxawmg6fqv7z35kbhocugdu47y/spack-src/src/H5VLcallback.c line 3675 in H5VL__file_open(): open failed
    major: Virtual Object Layer
    minor: Can't open object
  #004: /tmp/ialberto/spack-stage/spack-stage-hdf5-1.14.0-6nwv5jrxawmg6fqv7z35kbhocugdu47y/spack-src/src/H5VLnative_file.c line 128 in H5VL__native_file_open(): unable to open file
    major: File accessibility
    minor: Unable to open file
  #005: /tmp/ialberto/spack-stage/spack-stage-hdf5-1.14.0-6nwv5jrxawmg6fqv7z35kbhocugdu47y/spack-src/src/H5Fint.c line 1809 in H5F_open(): unable to open file: name = 'trid-ref.h5', tent_flags = 11
    major: File accessibility
    minor: Unable to open file
  #006: /tmp/ialberto/spack-stage/spack-stage-hdf5-1.14.0-6nwv5jrxawmg6fqv7z35kbhocugdu47y/spack-src/src/H5FD.c line 769 in H5FD_open(): open failed
    major: Virtual File Layer
    minor: Unable to initialize object
  #007: /tmp/ialberto/spack-stage/spack-stage-hdf5-1.14.0-6nwv5jrxawmg6fqv7z35kbhocugdu47y/spack-src/src/H5FDmpio.c line 907 in H5FD__mpio_open(): MPI_File_open failed: MPI error string is 'Read-only file or filesystem name, error stack:
ADIOI_CRAY_OPEN(111): Read access is required to this file'
    major: Internal error (too specific to document in detail)
    minor: Some MPI function failed
terminate called after throwing an instance of 'pika::detail::exception_with_info<pika::exception>'
  what():  attempted to call set_error on empty any_receiver: pika(bad_function_call)
srun: error: nid00008: task 0: Exited with exit code 255
srun: launch/slurm: _step_signal: Terminating StepId=47422047.0
```

And this on a RW filesyem, that I did just for double checking that it actually works as expected.
```bash
ialberto@daint106:/scratch/snx3000/ialberto> DLAF_DEBUG_DUMP_TRISOLVER_DATA=1 srun -n 1 -c72 ~/workspace/dla-future.hdf5/build-mc/miniapp/miniapp_eigensolver --matrix-size 10240 --block-size 512 --nwarmups=0 --nruns 1
[0]
[0] 62.6493s dL (10240, 10240) (512, 512) 128 (1, 1) 36 MC
```

If then you re-run it, since the file already exists, it aborts

```bash
ialberto@daint106:/scratch/snx3000/ialberto> DLAF_DEBUG_DUMP_TRISOLVER_DATA=1 srun -n 1 -c72 ~/workspace/dla-future.hdf5/build-mc/miniapp/miniapp_eigensolver --matrix-size 10240 --block-size 512 --nwarmups=0 --nruns 1
[0]
HDF5-DIAG: Error detected in HDF5 (1.14.0) MPI-process 0:
  #000: /tmp/ialberto/spack-stage/spack-stage-hdf5-1.14.0-6nwv5jrxawmg6fqv7z35kbhocugdu47y/spack-src/src/H5D.c line 186 in H5Dcreate2(): unable to synchronously create dataset
    major: Dataset
    minor: Unable to create file
  #001: /tmp/ialberto/spack-stage/spack-stage-hdf5-1.14.0-6nwv5jrxawmg6fqv7z35kbhocugdu47y/spack-src/src/H5D.c line 134 in H5D__create_api_common(): unable to create dataset
    major: Dataset
    minor: Unable to create file
  #002: /tmp/ialberto/spack-stage/spack-stage-hdf5-1.14.0-6nwv5jrxawmg6fqv7z35kbhocugdu47y/spack-src/src/H5VLcallback.c line 1876 in H5VL_dataset_create(): dataset create failed
    major: Virtual Object Layer
    minor: Unable to create file
  #003: /tmp/ialberto/spack-stage/spack-stage-hdf5-1.14.0-6nwv5jrxawmg6fqv7z35kbhocugdu47y/spack-src/src/H5VLcallback.c line 1841 in H5VL__dataset_create(): dataset create failed
    major: Virtual Object Layer
    minor: Unable to create file
  #004: /tmp/ialberto/spack-stage/spack-stage-hdf5-1.14.0-6nwv5jrxawmg6fqv7z35kbhocugdu47y/spack-src/src/H5VLnative_dataset.c line 269 in H5VL__native_dataset_create(): unable to create dataset
    major: Dataset
    minor: Unable to initialize object
  #005: /tmp/ialberto/spack-stage/spack-stage-hdf5-1.14.0-6nwv5jrxawmg6fqv7z35kbhocugdu47y/spack-src/src/H5Dint.c line 349 in H5D__create_named(): unable to create and link to dataset
    major: Dataset
    minor: Unable to initialize object
  #006: /tmp/ialberto/spack-stage/spack-stage-hdf5-1.14.0-6nwv5jrxawmg6fqv7z35kbhocugdu47y/spack-src/src/H5Lint.c line 516 in H5L_link_object(): unable to create new link to object
    major: Links
    minor: Unable to initialize object
  #007: /tmp/ialberto/spack-stage/spack-stage-hdf5-1.14.0-6nwv5jrxawmg6fqv7z35kbhocugdu47y/spack-src/src/H5Lint.c line 759 in H5L__create_real(): can't insert link
    major: Links
    minor: Unable to insert object
  #008: /tmp/ialberto/spack-stage/spack-stage-hdf5-1.14.0-6nwv5jrxawmg6fqv7z35kbhocugdu47y/spack-src/src/H5Gtraverse.c line 836 in H5G_traverse(): internal path traversal failed
    major: Symbol table
    minor: Object not found
  #009: /tmp/ialberto/spack-stage/spack-stage-hdf5-1.14.0-6nwv5jrxawmg6fqv7z35kbhocugdu47y/spack-src/src/H5Gtraverse.c line 613 in H5G__traverse_real(): traversal operator failed
    major: Symbol table
    minor: Callback failed
  #010: /tmp/ialberto/spack-stage/spack-stage-hdf5-1.14.0-6nwv5jrxawmg6fqv7z35kbhocugdu47y/spack-src/src/H5Lint.c line 551 in H5L__link_cb(): name already exists
    major: Links
    minor: Object already exists
srun: error: nid00008: task 0: Segmentation fault (core dumped)
srun: launch/slurm: _step_signal: Terminating StepId=47422047.2
```

Reply:
Awesome, thanks! Those are not perfect, but better than I expected. I've seen some other applications only report:
```
srun: error: nid00008: task 0: Segmentation fault (core dumped)
```
in similar cases which is why I asked. This is far better than that.